### PR TITLE
Integrate ref count module into local memory store

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

Instead of removing objects from the memory store on get, allow the ref count module to manage the object lifetimes. This is flagged on by default.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: src/ray/core_worker/reference_count.h [hunk 1]
```diff
@@ -37,6 +37,12 @@ class ReferenceCounter {
   /// Returns the total number of ObjectIDs currently in scope.
   size_t NumObjectIDsInScope() const LOCKS_EXCLUDED(mutex_);
 
+  /// Return whether this object has an active reference.
+  bool HasReference(const ObjectID &object_id) LOCKS_EXCLUDED(mutex_);
+
+  /// Sets or resets the callback to run when an object is deleted.
+  void OnObjectDeleted(std::function<void(const ObjectID &)> callback);
```
Comment (stephanie-wang) [inline src/ray/core_worker/reference_count.h]: Can we just pass this into the constructor? It seems like the CoreWorker could set the callback when constructing ref_counter_.

File: src/ray/core_worker/store_provider/memory_store/memory_store.cc [hunk 2]
```diff
@@ -104,10 +104,20 @@ std::shared_ptr<RayObject> GetRequest::Get(const ObjectID &object_id) const {
...
+  if (counter != nullptr) {
+    ref_counter_ = counter;
+    ref_counter_->OnObjectDeleted([this](const ObjectID &obj_id) { Delete({obj_id}); });
+  }
+}
 
 Status CoreWorkerMemoryStore::Put(const ObjectID &object_id, const RayObject &object) {
   std::unique_lock<std::mutex> lock(lock_);
+
+  if (ref_counter_ && !ref_counter_->HasReference(object_id)) {
```
Comment (stephanie-wang) [inline src/ray/core_worker/store_provider/memory_store/memory_store.cc]: Ahh is this an optimization for if the return value is out of scope? Nice. Can you add a comment?

**Review by stephanie-wang — APPROVED** [review]

File: src/ray/core_worker/reference_count.cc [hunk 3]
```diff
@@ -56,9 +56,22 @@ void ReferenceCounter::RemoveReferenceRecursive(const ObjectID &object_id) {
...
+      on_delete_(object_id);
+    }
   }
 }
 
+bool ReferenceCounter::HasReference(const ObjectID &object_id) {
+  absl::MutexLock lock(&mutex_);
+  return object_id_refs_.find(object_id) != object_id_refs_.end();
+}
+
```
Comment (edoakes) [inline src/ray/core_worker/reference_count.cc]: Do we set this dynamically because we initialize the local memory store after the ref counter? If so this should probably append a callback rather than overwrite an existing one if it exists as we will want to perform multiple operations when an object id goes out of scope. That or set one callback into a core worker "HandleObjectOutOfScope" when we construct this.

File: src/ray/core_worker/store_provider/memory_store/memory_store.cc (hunk 2 above)
Comment (ericl) [reply src/ray/core_worker/store_provider/memory_store/memory_store.cc]: Done

File: src/ray/core_worker/reference_count.h (hunk 1 above)
Comment (ericl) [reply src/ray/core_worker/reference_count.h]: Not easily, since it's a circular dependency. I think it's more straightforward to mutate the class.

File: src/ray/core_worker/reference_count.cc (hunk 3 above)
Comment (ericl) [reply src/ray/core_worker/reference_count.cc]: I added a RAY_CHECK that this is only called once, to avoid any correctness concerns.

Comment (ericl) [reply src/ray/core_worker/reference_count.cc]: > will want to perform multiple operations

we can add that later.

File: src/ray/core_worker/reference_count.h (hunk 1 above)
Comment (stephanie-wang) [reply src/ray/core_worker/reference_count.h]: Can't you pass in `[this](...) { memory_store_.Delete(...) }` from the core worker constructor?

Comment (ericl) [reply src/ray/core_worker/reference_count.h]: I thought about that, but it seems more subtle to me. The current approach is straightforward, works, and is already implemented.

File: src/ray/core_worker/store_provider/memory_store/memory_store.cc (hunk 2 above)
Comment (zhijunfu) [reply src/ray/core_worker/store_provider/memory_store/memory_store.cc]: @ericl I might miss something here - if an object ID is already out-of-scope, how would user code invoke a Put in this case?

File: src/ray/core_worker/core_worker.cc [hunk 4]
```diff
@@ -72,7 +72,9 @@ CoreWorker::CoreWorker(const WorkerType worker_type, const Language language,
       heartbeat_timer_(io_service_),
       worker_server_(WorkerTypeString(worker_type), 0 /* let grpc choose a port */),
       gcs_client_(gcs_options),
-      memory_store_(std::make_shared<CoreWorkerMemoryStore>()),
+      reference_counter_(std::make_shared<ReferenceCounter>()),
+      memory_store_(std::make_shared<CoreWorkerMemoryStore>(
+          RayConfig::instance().ref_counting_enabled() ? reference_counter_ : nullptr)),
```
Comment (zhijunfu) [inline src/ray/core_worker/core_worker.cc]: Could you enable this only for python workers? aka, only when worker language is python.

Because java doesn't implement ref counting yet, enabling this for java would cause memory leak in this case.

**Comment by ericl** [comment]
This seems to hang core worker tests, probably due to some edge case not handled.

File: src/ray/core_worker/core_worker.cc (hunk 4 above)
Comment (ericl) [reply src/ray/core_worker/core_worker.cc]: Thanks, fixed this.

File: src/ray/core_worker/store_provider/memory_store/memory_store.cc [hunk 5]
```diff
@@ -198,8 +205,11 @@ Status CoreWorkerMemoryStore::Get(const std::vector<ObjectID> &object_ids,
       }
     }
 
-    for (const auto &object_id : ids_to_remove) {
-      objects_.erase(object_id);
+    // Clean up the objects if ref counting is off.
```
Comment (edoakes) [inline src/ray/core_worker/store_provider/memory_store/memory_store.cc]: Can you add a `TODO` somewhere about removing this logic once we have GC in Java.

**Review by edoakes — APPROVED** [review]

**Comment by ericl** [comment]
I figured out why the test was hanging on Python2: it was because we didn't call SetDependencies for direct call tasks. Hence, the return id wasn't always referenced before the call returned (which discards the result if there isn't at least a 0-ref for the objid).

File: src/ray/core_worker/reference_count.h [hunk 6]
```diff
@@ -37,6 +41,9 @@ class ReferenceCounter {
   /// Returns the total number of ObjectIDs currently in scope.
   size_t NumObjectIDsInScope() const LOCKS_EXCLUDED(mutex_);
 
+  /// Return whether this object has an active reference.
```
[the remark this answers is not in the document]
Comment (ericl) [reply src/ray/core_worker/reference_count.h]: Done

File: src/ray/core_worker/store_provider/memory_store/memory_store.cc (hunk 5 above)
Comment (ericl) [reply src/ray/core_worker/store_provider/memory_store/memory_store.cc]: Done
